src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs

```diff
@@ -3021,32 +3021,48 @@ static T VectorizedSmall8(ref T xRef, ref T yRef, nuint remainder)
             BitOperations.TrailingZeroCount(mask.ExtractMostSignificantBits());
 
         /// <summary>Performs an element-wise operation on <paramref name="x"/> and writes the results to <paramref name="destination"/>.</summary>
-        /// <typeparam name="T">The element type.</typeparam>
+        /// <typeparam name="T">The element input type.</typeparam>
         /// <typeparam name="TUnaryOperator">Specifies the operation to perform on each element loaded from <paramref name="x"/>.</typeparam>
         private static void InvokeSpanIntoSpan<T, TUnaryOperator>(
             ReadOnlySpan<T> x, Span<T> destination)
-            where TUnaryOperator : struct, IUnaryOperator<T>
+            where TUnaryOperator : struct, IUnaryOperator<T, T> =>
+            InvokeSpanIntoSpan<T, T, TUnaryOperator>(x, destination);
+
+        /// <summary>Performs an element-wise operation on <paramref name="x"/> and writes the results to <paramref name="destination"/>.</summary>
+        /// <typeparam name="TInput">The element input type.</typeparam>
+        /// <typeparam name="TOutput">The element output type. Must be the same size as TInput if TInput and TOutput both support vectorization.</typeparam>
+        /// <typeparam name="TUnaryOperator">Specifies the operation to perform on each element loaded from <paramref name="x"/>.</typeparam>
+        /// <remarks>
+        /// This supports vectorizing the operation if <typeparamref name="TInput"/> and <typeparamref name="TOutput"/> are the same size.
+        /// Otherwise, it'll fall back to scalar operations.
+        /// </remarks>
+        private static void InvokeSpanIntoSpan<TInput, TOutput, TUnaryOperator>(
+            ReadOnlySpan<TInput> x, Span<TOutput> destination)
+            where TUnaryOperator : struct, IUnaryOperator<TInput, TOutput>
         {
             if (x.Length > destination.Length)
             {
                 ThrowHelper.ThrowArgument_DestinationTooShort();
             }
 
-            ValidateInputOutputSpanNonOverlapping(x, destination);
+            if (typeof(TInput) == typeof(TOutput))
+            {
+                ValidateInputOutputSpanNonOverlapping(x, Rename<TOutput, TInput>(destination));
+            }
```

Comment:
We may want this check "always" as users can (and potentially will) alias their spans in some cases.

Of course, we could decide to leave that as "here be dragons", but a comment documenting that is likely worthwhile in that case.